openfl/experimental/workflow/component/director/director.py

```diff
@@ -0,0 +1,314 @@
+                root certificate for TLS. Defaults to None.
+            private_key (Optional[Union[Path, str]]): The path to the private
+                key for TLS. Defaults to None.
+            certificate (Optional[Union[Path, str]]): The path to the
+                certificate for TLS. Defaults to None.
+            director_config (Optional[Path]): Path to director_config file
+            envoy_health_check_period (int): The period for health check of envoys
+            in seconds.
+            install_requirements (bool, optional): A flag indicating if the
+                requirements should be installed. Defaults to True.
+        """
+        self.tls = tls
+        self.root_certificate = root_certificate
+        self.private_key = private_key
+        self.certificate = certificate
+        self.director_config = director_config
+        self.install_requirements = install_requirements
+        self._flow_status = asyncio.Queue()
+
+        self.experiments_registry = ExperimentsRegistry()
+        self.col_exp = {}
+        self.col_exp_queues = defaultdict(asyncio.Queue)
+        self._envoy_registry = {}
+        self.envoy_health_check_period = envoy_health_check_period
+        # authorized_cols refers to envoy & collaborator pair (one to one mapping)
+        self.authorized_cols = []
+
+    async def start_experiment_execution_loop(self) -> None:
+        """Run tasks and experiments here"""
+        loop = asyncio.get_event_loop()
+        while True:
+            try:
+                async with self.experiments_registry.get_next_experiment() as experiment:
+                    await self._wait_for_authorized_envoys()
+                    run_aggregator_future = loop.create_task(
+                        experiment.start(
+                            root_certificate=self.root_certificate,
+                            certificate=self.certificate,
+                            private_key=self.private_key,
+                            tls=self.tls,
+                            director_config=self.director_config,
+                            install_requirements=False,
+                        )
+                    )
+                    # Adding the experiment to collaborators queues
+                    for col_name in experiment.collaborators:
+                        queue = self.col_exp_queues[col_name]
+                        await queue.put(experiment.name)
+                    # Wait for the experiment to complete and save the result
+                    flow_status = await run_aggregator_future
+                    await self._flow_status.put(flow_status)
+            except Exception as e:
+                logger.error(f"Error while executing experiment: {e}")
+                raise
+
+    async def _wait_for_authorized_envoys(self) -> None:
+        """Wait until all authorized envoys are connected"""
+        while not all(envoy in self.get_envoys().keys() for envoy in self.authorized_cols):
```

Comment:
as per my comment above, wouldn't this "read" better as `... for envoy in self.authorized_envoys):` ?

Reply:
As discussed there is one to one mapping between Envoys and collaborators, and to maintain consistency in
code `self.authorized_cols` is currently used. Marking this as resolved following the internal discussion with Teo.